host/crypto/bcrypt/rsa.c

```diff
@@ -9,446 +10,325 @@
-    {
-        NTSTATUS status = BCryptImportKeyPair(
-            BCRYPT_RSA_ALG_HANDLE,
-            NULL,
-            BCRYPT_RSAPRIVATE_BLOB,
-            &handle,
-            rsa_blob,
-            rsa_blob_size,
-            0);
-
-        if (!BCRYPT_SUCCESS(status))
-            OE_RAISE(OE_CRYPTO_ERROR);
-    }
-
-    /* Step 4: Convery BCrypt Handle to OE type. */
-    {
-        oe_private_key_t* pkey = (oe_private_key_t*)private_key;
-        pkey->magic = OE_RSA_PRIVATE_KEY_MAGIC;
-        pkey->pkey = handle;
-        handle = NULL;
+            OE_RAISE_MSG(
+                OE_CRYPTO_ERROR,
+                "CryptEncodeObjectEx failed (err=%#x)\n",
+                GetLastError());
     }
 
     result = OE_OK;
 
 done:
-    if (handle)
-        BCryptDestroyKey(handle);
-
-    /*
-     * Make sure to zero out all confidential (private key) data.
-     * Note that rsa_blob must be freed with LocalFree and before der_data
-     * due to constraints in CryptObjectDecodeEx.
-     */
-    if (rsa_blob)
+    if (key_blob)
     {
-        oe_secure_zero_fill(rsa_blob, rsa_blob_size);
-        LocalFree(rsa_blob);
-        rsa_blob_size = 0;
-    }
-
-    if (der_data)
-    {
-        oe_secure_zero_fill(der_data, der_size);
-        free(der_data);
-        der_size = 0;
+        oe_secure_zero_fill(key_blob, key_blob_size);
+        free(key_blob);
+        key_blob_size = 0;
     }
 
     return result;
 }
 
-oe_result_t oe_rsa_private_key_free(oe_rsa_private_key_t* private_key)
+/* Caller is responsible for calling free on padding_info->config */
+static oe_result_t _get_padding_info(
+    oe_hash_type_t hash_type,
+    size_t hash_size,
+    oe_bcrypt_padding_info_t* padding_info)
 {
     oe_result_t result = OE_UNEXPECTED;
+    PCWSTR hash_algorithm = NULL;
 
-    if (private_key)
+    /* Check for support hash types and correct sizes. */
+    switch (hash_type)
     {
-        oe_private_key_t* impl = (oe_private_key_t*)private_key;
-
-        /* Check parameter */
-        if (!oe_private_key_is_valid(impl))
+        case OE_HASH_TYPE_SHA256:
+            if (hash_size != 32)
+                OE_RAISE(OE_INVALID_PARAMETER);
+            hash_algorithm = BCRYPT_SHA256_ALGORITHM;
+            break;
+        case OE_HASH_TYPE_SHA512:
+            if (hash_size != 64)
+                OE_RAISE(OE_INVALID_PARAMETER);
+            hash_algorithm = BCRYPT_SHA512_ALGORITHM;
+            break;
+        default:
             OE_RAISE(OE_INVALID_PARAMETER);
-
-        /* Release the key */
-        BCryptDestroyKey(impl->pkey);
-
-        /* Clear the fields of the implementation */
-        oe_secure_zero_fill(impl, sizeof(*impl));
     }
 
+    /*
+     * Note that we use the less secure PKCS1 signature padding
+     * because Intel requires it for SGX enclave signatures.
+     */
+    padding_info->type = BCRYPT_PAD_PKCS1;
+    padding_info->config = malloc(sizeof(BCRYPT_PKCS1_PADDING_INFO));
```

Comment:
Check of out of memory here?

Reply:
Thanks for the catch, added allocation check.